File: lib/apps/triplet_loader.dart
import 'dart:io';
import 'dart:math';
import 'package:image/image.dart' as img;
import 'package:dart_cuda/gpu_tensor.dart';

class TripletLoader {
  final String rootPath;
  final int imageSize;
  final Map<String, List<File>> _identityMap = {};
  final Random _random = Random();

  TripletLoader(this.rootPath, this.imageSize) {
    _scanDataset();
  }

  void _scanDataset() {
    final rootDir = Directory(rootPath);
    for (var entity in rootDir.listSync()) {
      if (entity is Directory) {
        final personName = entity.path.split(Platform.pathSeparator).last;
        _identityMap[personName] = entity
            .listSync()
            .whereType<File>()
            .where((f) => f.path.endsWith('.jpg') || f.path.endsWith('.png'))
            .toList();
      }
    }
  }

  /// Loads a batch of [batchSize] triplets directly into GPU memory
  Map<String, Tensor> nextBatch(int batchSize, List<Tensor> tracker) {
    List<double> anchors = [];
    List<double> positives = [];
    List<double> negatives = [];

    for (int i = 0; i < batchSize; i++) {
      // 1. Pick a random person for Anchor & Positive
      final people = _identityMap.keys.toList();
      final personA = people[_random.nextInt(people.length)];
      final imagesA = _identityMap[personA]!;

      final imgIdx1 = _random.nextInt(imagesA.length);
      int imgIdx2;
      do {
        imgIdx2 = _random.nextInt(imagesA.length);
      } while (imgIdx1 == imgIdx2 && imagesA.length > 1);

      // 2. Pick a different person for Negative
      String personB;
      do {
        personB = people[_random.nextInt(people.length)];
      } while (personA == personB);
      final imagesB = _identityMap[personB]!;

      // 3. Process and Flatten
      anchors.addAll(_processImage(imagesA[imgIdx1]));
      positives.addAll(_processImage(imagesA[imgIdx2]));
      negatives.addAll(_processImage(imagesB[_random.nextInt(imagesB.length)]));
    }

    final anchorTensor = Tensor.fromList([
      batchSize,
      imageSize * imageSize * 3,
    ], anchors);
    final positiveTensor = Tensor.fromList([
      batchSize,
      imageSize * imageSize * 3,
    ], positives);
    final negativeTensor = Tensor.fromList([
      batchSize,
      imageSize * imageSize * 3,
    ], negatives);

    tracker.addAll([anchorTensor, positiveTensor, negativeTensor]);

    return {
      'anchor': anchorTensor,
      'positive': positiveTensor,
      'negative': negativeTensor,
    };
  }

  List<double> _processImage(File file) {
    final raw = img.decodeImage(file.readAsBytesSync());
    final resized = img.copyResize(raw!, width: imageSize, height: imageSize);
    // Convert to normalized Float32 values (0.0 to 1.0)
    return resized
        .map((p) => [p.r / 255.0, p.g / 255.0, p.b / 255.0])
        .expand((e) => e)
        .toList();
  }
}

File: python/dart_cuda/loaders/triplet_loader.py
"""Triplet loader — mirrors `lib/loaders/triplet_loader.dart`.

Requires Pillow.
"""

from __future__ import annotations

import os
import random
from typing import Dict, List

from ..core.tensor.gpu_tensor import Tensor


def _require_pil():
    try:
        from PIL import Image  # noqa: F401

        return Image
    except ImportError as exc:  # pragma: no cover
        raise ImportError("Pillow is required: pip install Pillow") from exc


class TripletLoader:
    def __init__(self, root_path: str, image_size: int):
        self.root_path = root_path
        self.image_size = image_size
        self._identity_map: Dict[str, List[str]] = {}
        self._random = random.Random()
        self._scan_dataset()

    def _scan_dataset(self) -> None:
        for name in os.listdir(self.root_path):
            full = os.path.join(self.root_path, name)
            if os.path.isdir(full):
                files = [
                    os.path.join(full, f)
                    for f in os.listdir(full)
                    if f.lower().endswith((".jpg", ".jpeg", ".png"))
                ]
                if files:
                    self._identity_map[name] = files

    def _process_image(self, file_path: str) -> List[float]:
        Image = _require_pil()
        img = (
            Image.open(file_path)
            .convert("RGB")
            .resize((self.image_size, self.image_size))
        )
        out: List[float] = []
        for r, g, b in img.getdata():
            out.append(r / 255.0)
            out.append(g / 255.0)
            out.append(b / 255.0)
        return out

    def next_batch(self, batch_size: int) -> Dict[str, Tensor]:
        anchors: List[float] = []
        positives: List[float] = []
        negatives: List[float] = []
        people = list(self._identity_map.keys())
        if len(people) < 2:
            raise RuntimeError("Need at least 2 identities for triplets")

        for _ in range(batch_size):
            person_a = people[self._random.randrange(len(people))]
            images_a = self._identity_map[person_a]
            i1 = self._random.randrange(len(images_a))
            i2 = self._random.randrange(len(images_a))
            if len(images_a) > 1:
                while i1 == i2:
                    i2 = self._random.randrange(len(images_a))
            person_b = person_a
            while person_b == person_a:
                person_b = people[self._random.randrange(len(people))]
            images_b = self._identity_map[person_b]

            anchors.extend(self._process_image(images_a[i1]))
            positives.extend(self._process_image(images_a[i2]))
            negatives.extend(
                self._process_image(images_b[self._random.randrange(len(images_b))])
            )

        feat = self.image_size * self.image_size * 3
        return {
            "anchor": Tensor.from_list([batch_size, feat], anchors),
            "positive": Tensor.from_list([batch_size, feat], positives),
            "negative": Tensor.from_list([batch_size, feat], negatives),
        }

    nextBatch = next_batch
